crypto/getenv.c

```diff
@@ -15,6 +15,48 @@
 #include "internal/cryptlib.h"
 #include "internal/e_os.h"
 
+#if defined (OPENSSL_NETCAP_ALLOW_ENV) && defined(__linux__)
+# include "internal/nelem.h"
+#endif
+
+/*
+ * Additional checks for securely allowing access to selected environment variables
+ */
+#if defined (OPENSSL_NETCAP_ALLOW_ENV) && defined(__linux__)
+
+/*
+ * Utility function to check if a string is in an array of strings. returns true if found
+ */
+int str_in_list(const char *candidate, const char *list[], int n)
+{
+
+  if (candidate == NULL || list == NULL || n < 1)
+    return 0;
+
+  int i;
+  for ( i=0; i<n; i++) {
+    if (strcmp(candidate, list[i]) == 0)
+      return 1;
+  }
+  return 0;
+}
+
+char *restricted_getenv(const char *name)
+{
+    /* Environment variables to permit in secure mode (subject to conditions). Extend as needed */
+    const char *permitted_env[] = { "OPENSSL_MODULES", "OPENSSL_CONF" };
```

Comment:
This seems like a very bad idea.  Those are the two variables that allow you to specify what code to load by default (via provider loading and activation) and give you the means to redirect this to a place where you can put your code.  So this makes OpenSSL kind of insecure in this case.

Reply:
I understand the concern -- putting aside the questions around configurability,  the core proposal is specific to net-cap-bind - indeed it should only be a change where this is the ONE and ONLY capability set. Any more, and the test fails, and we use secure_getenv(). (I added a PR after the initial issue discussion so as to be explicit about what was proposed)

As you say, it does allow for injection of the providers (by spec of directory, and configuration) - which is where the risk of malicious code lies.

This capability gives the ability for a process to bind to a port < 1024, nothing else. So the risk is of injected code becoming active on those ports. This is the core question - is that reasonable?

The 'node' project made a similar change in https://github.com/nodejs/node/pull/37727 - of course that's a different project, with a different role

Reply:
There is no point in allowing just these two variables and routing the rest through secure_getenv(). Just use this for any ossl_safe_getenv() calls.

And of course all this is behing  OPENSSL_NETCAP_ALLOW_ENV ifdef so this is not built by default. So IMO ok.

Reply:
@juergenchrist I am certainly concerned about avoiding opening up security holes.

The core proposal was focussed specifically on net-cap-bind capability, which allows a process to bind to a port <1024. If any other capabilities are found, this new logic should not apply, regardless of net-cap-bind being set,and environment variables blocked .

The potential exposure with this capability would be to allow malicious or unexpected code to become active on a port <1024 where previously they could not, in the case where an admin had granted the capability, but does not control the values of the environment ie provider, libraries etc. Today the combination of binding to ports < 1024 & having control of the configuration through variables is not possible by a non-admin user.

The 'node' project opted to permit the loosening of the check, albeit that is one specific application. I created the PR to be explicit about the proposal, to see if the community would find this restricted loosening  appropriate. The discussion above also has touched on config, but it's this specific case where I see this level of appropriateness.

The overall intent is to enable the use of new quantum safe encryption algorithms (ie as provided by the liboqs project and liboqs openssl provider) to be easier. This can be done via admin configuration of openssl.cnf/placing libraries in appropriate location,  via the app utilizing openssl, or (sometimes fiddly) wrapping techniques. We're trying to see if we can avoid those steps.

Long term perhaps these new algorithms will make it into core openssl, at which point maybe issue won't apply, but even then crypto-agility is important, so it will then just be a new set of quantum-safe algorithms.... !